File: backend/src/main/java/com/groupone/backend/features/writing/dto/EssaySubmissionResponse.java
package com.groupone.backend.features.writing.dto;

import com.groupone.backend.features.writing.GradingType;
import com.groupone.backend.features.writing.WritingSubmission;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.time.LocalDateTime;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class EssaySubmissionResponse {
    private Long id;
    private TopicResponse topic;
    private String content;
    private GradingType gradingType;
    private String aiFeedback;
    private Double score;
    private LocalDateTime createdAt;
    private String status;

    
    private Double taskAchievement;
    private Double cohesionCoherence;
    private Double lexicalResource;
    private Double grammaticalRange;
    private String teacherFeedback;
    private String taskAchievementReason;
    private String cohesionCoherenceReason;
    private String lexicalResourceReason;
    private String grammaticalRangeReason;
    private String correctionsJson;

    public static EssaySubmissionResponse fromEntity(WritingSubmission submission) {
        return EssaySubmissionResponse.builder()
                .id(submission.getId())
                .topic(TopicResponse.fromEntity(submission.getQuestion()))
                .content(submission.getContent())
                .gradingType(submission.getGradingType())
                .aiFeedback(submission.getAiFeedback())
                .score(submission.getScore())
                .createdAt(submission.getCreatedAt())
                .status(submission.getStatus() != null ? submission.getStatus().name() : null)
                .taskAchievement(submission.getTaskAchievement())
                .cohesionCoherence(submission.getCohesionCoherence())
                .lexicalResource(submission.getLexicalResource())
                .grammaticalRange(submission.getGrammaticalRange())
                .teacherFeedback(submission.getTeacherFeedback())
                .taskAchievementReason(submission.getTaskAchievementReason())
                .cohesionCoherenceReason(submission.getCohesionCoherenceReason())
                .lexicalResourceReason(submission.getLexicalResourceReason())
                .grammaticalRangeReason(submission.getGrammaticalRangeReason())
                .correctionsJson(submission.getCorrectionsJson())
                .build();
    }
}

File: mobile-desktop/lib/features/study_sections/writing/models/essay_submission_response.dart
import 'topic_model.dart';
import 'writing_correction.dart';

class EssaySubmissionResponse {
  final int id;
  final Topic topic;
  final String content;
  final String gradingType;
  final String? aiFeedback;
  final double? score;
  final String? status;
  final String? createdAt;

  final double? taskAchievement;
  final double? grammaticalRange;
  final double? lexicalResource;
  final double? cohesionCoherence;
  final String? teacherFeedback;
  final String? taskAchievementReason;
  final String? cohesionCoherenceReason;
  final String? lexicalResourceReason;
  final String? grammaticalRangeReason;
  final List<WritingCorrection> corrections;

  EssaySubmissionResponse({
    required this.id,
    required this.topic,
    required this.content,
    required this.gradingType,
    this.aiFeedback,
    this.score,
    this.status,
    this.createdAt,
    this.taskAchievement,
    this.grammaticalRange,
    this.lexicalResource,
    this.cohesionCoherence,
    this.teacherFeedback,
    this.taskAchievementReason,
    this.cohesionCoherenceReason,
    this.lexicalResourceReason,
    this.grammaticalRangeReason,
    this.corrections = const [],
  });

  factory EssaySubmissionResponse.fromJson(Map<String, dynamic> json) {
    var correctionsList = <WritingCorrection>[];
    if (json['corrections'] != null) {
      correctionsList = (json['corrections'] as List)
          .map((i) => WritingCorrection.fromJson(i))
          .toList();
    }

    return EssaySubmissionResponse(
      id: json['id'],
      topic: Topic.fromJson(json['topic']),
      content: json['content'],
       gradingType: json['gradingType'],
      aiFeedback: json['aiFeedback'],
      score: json['score'] != null ? (json['score'] as num).toDouble() : null,
      status: json['status'],
      createdAt: json['createdAt'],
      taskAchievement: json['taskAchievement'] != null
          ? (json['taskAchievement'] as num).toDouble()
          : null,
      grammaticalRange: json['grammaticalRange'] != null
          ? (json['grammaticalRange'] as num).toDouble()
          : null,
      lexicalResource: json['lexicalResource'] != null
          ? (json['lexicalResource'] as num).toDouble()
          : null,
      cohesionCoherence: json['cohesionCoherence'] != null
          ? (json['cohesionCoherence'] as num).toDouble()
          : null,
      teacherFeedback: json['teacherFeedback'],
      taskAchievementReason: json['taskAchievementReason'],
      cohesionCoherenceReason: json['cohesionCoherenceReason'],
      lexicalResourceReason: json['lexicalResourceReason'],
      grammaticalRangeReason: json['grammaticalRangeReason'],
      corrections: correctionsList,
    );
  }
}
